[WIP] Fix off-by-one error resulted from memory methods checked_offset & checked_range_offset incorrect usage

iulianbarbu opened the pull request:
This PR get rid of some discovered incorrect usages of `checked_offset` and `checked_range_offset` from memory_model crate. It also renames memory model
functions like `region_end` and `end_addr` which with their naming bring ambiguity
in how their usage should be.

Signed-off-by: Iulian Barbu <iul@amazon.com>

## Reason for This PR

#1296 

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria. Where there are two options, keep one.]`  
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [ ] Either this PR is linked to an issue, or, the reason for this PR is
      clearly provided. 
- [ ] The description of changes is clear and encompassing.
- [ ] Either no docs need to be updated as part of this PR, or, the required
      doc changes are included in this PR. Docs in scope are all `*.md` files
      located either in the repository root, or in the `docs/` directory.
- [ ] Either no code has been touched, or, code-level documentation for touched
      code is included in this PR.
- [ ] Either no API changes are included in this PR, or, the API changes are
      reflected in `firecracker/swagger.yaml`.
- [ ] Either the changes in this PR have no user impact, or, the changes in
      this PR have user impact and have been added to the `CHANGELOG.md` file.

acatangiu commented:
Haven't yet looked at the code, but wanted to call-out the need for a better commit (and PR) commit summary (the first line - the title). It should summarize the intention of the changes along with the area/module they are in.

iulianbarbu commented:
> Haven't yet looked at the code, but wanted to call-out the need for a better commit (and PR) summary (the first line - the title). It should summarize the intention of the changes along with the area/module they are in.

I've updated accordingly. Please take a look at the commit title again. I've added the modules where the changes reside in the commit description, because of the 50 chars length rule for the commit title.

dhrgit commented on devices/src/virtio/queue.rs:
```diff
@@ -105,12 +106,30 @@ impl<'a> DescriptorChain<'a> {
+        let last_addr_opt = upper_limit_opt.unwrap().checked_sub(1);
+        if last_addr_opt.is_none() {
+            return false;
+        }
```
I don't really understand this one either. `x.checked_sub(1)` will only fail for `x == 0`, so what is that you are checking here?

dhrgit commented on memory_model/src/guest_memory.rs:
```diff
@@ -124,13 +125,11 @@ impl GuestMemory {
 
     /// Returns the address plus the offset if it is in range.
     pub fn checked_offset(&self, base: GuestAddress, offset: usize) -> Option<GuestAddress> {
```
Cleaning this up is great. How about, while at it, also fixing `region_end()`. It has a misleading name, since the returned value is not actually an end (i.e. the last usable byte, `base + len - 1`), but the first non-usable byte (i.e. `base + len`).

dhrgit replied:
I favor having `region_end()` actually return the region end (i.e. last usable byte). That's because `base + end - 1` is always a valid number (when both `base` and `end` are valid), as opposed to `base + end`, which may overflow for valid values of `base` and `end`.

iulianbarbu replied:
I agree. I will update region_end() as you suggest.

iulianbarbu commented on devices/src/virtio/vsock/epoll_handler.rs:
```diff
@@ -367,20 +351,6 @@ mod tests {
-            assert_eq!(ctx.handler.process_rx(), true);
-            assert_eq!(ctx.guest_rxvq.used.idx.get(), 1);
-            assert_eq!(ctx.handler.backend.rx_ok_cnt, 0);
-        }
```
The same applies also for RX.

iulianbarbu commented on devices/src/virtio/vsock/epoll_handler.rs:
```diff
@@ -306,22 +306,6 @@ mod tests {
-            // reached the backend.
-            assert_eq!(ctx.guest_txvq.used.idx.get(), 1);
-            assert_eq!(ctx.handler.backend.tx_ok_cnt, 0);
-        }
```
I will add the malformed descriptors tests, for TX and RX, but first, we have to differentiate between invalid descriptors virtio queue wise, and invalid descriptors device wise. 

Example:
- an invalid descriptor virtio queue wise is a descriptor which points to a memory region of length 0.
- an invalid descriptor device wise, for vsock device lets say, is a header_descriptor which points to a memory region with a length lower than VSOCK_PKT_HDR_SIZE.

As you can see, for these two types of invalid descriptors, the overlapping case is when header_descriptor memory region length is 0. This happens in the malformed descriptors tests for RX and TX, and should be modified accordingly, to what makes sense to be an invalid descriptor for the vsock device, and more general, for a specific device.

raduiliescu commented:
In your last push the queue unit test is failing- test virtio::queue::tests::test_checked_new_descriptor_chain ... FAILED
Event though is WIP at least unit tests should be clean at every push.

iulianbarbu commented:
> In your last push the queue unit test is failing- test virtio::queue::tests::test_checked_new_descriptor_chain ... FAILED
> Event though is WIP at least unit tests should be clean at every push.

Yes, I get your point. I was thinking more on actual backing up my latest changes. I can also keep a mirror branch, to push to it, which is not linked in any way to the PR branch, but it is not that straight forward. I guess it is not so impactful if we force push changes that are on their way to be fixed, or is it?

petreeftime commented:
> In your last push the queue unit test is failing- test virtio::queue::tests::test_checked_new_descriptor_chain ... FAILED
> Event though is WIP at least unit tests should be clean at every push.

As long as it doesn't create unnecessary work for other people, why would this be an issue?
